### PR TITLE
ENH: Add Neon SIMD implementations for add, sub, mul, and div

### PR DESCRIPTION
**Pull request by DumbMice**
## Background:
There have already been ASV and SSE2 SIMD implementations for add, subtract, multiply and divide, and this is a reimplementation of existing SIMD codes for **float32/64** in neon intrinsics in order to add supports on ARM platform. Resultantly, optimisation reduces the computation time by a percentage of 40%-70% on this specific ARM machine.
## System Info:

   * HardWare: KunPeng
   * Processor: ARMv8 2.6GMHZ 8 processors
   * OS: Linux ecs-9d50 4.19.36-vhulk1905.1.0.h276.eulerosv2r8.aarch64
## Benchmark
  |    before       |   after             |ratio|    bench|
  |    ---          |   ---               | --- |    --- |
  |        197±2μs  |      117±0.8μs      | 0.59|  bench_ufunc.CustomInplace.time_double_add_temp|
  |    51.6±0.03μs  |    28.8±0.06μs      | 0.56|  bench_ufunc.CustomScalar.time_divide_scalar2_inplace(<class 'numpy.float32'>)|
  |    50.5±0.06μs  |    27.3±0.03μs      | 0.54|  bench_ufunc.CustomScalar.time_divide_scalar2(<class 'numpy.float32'>)|
  |    24.5±0.07μs  |    13.0±0.03μs      | 0.53|  bench_ufunc.CustomScalar.time_add_scalar2(<class 'numpy.float64'>)|
  |        175±8μs  |     91.1±0.7μs      | 0.52|  bench_ufunc.CustomInplace.time_double_add|
  |        346±2μs  |      124±0.7μs      | 0.36|  bench_ufunc.CustomInplace.time_float_add_temp|
  |     24.5±0.4μs  |    8.32±0.04μs      | 0.34|  bench_ufunc.CustomScalar.time_add_scalar2(<class 'numpy.float32'>)|
  |        321±1μs  |     96.2±0.6μs      | 0.30|  bench_ufunc.CustomInplace.time_float_add|

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

## Note
I have noticed a recent tendency of replacing newly added SIMD intrinsics of existing implementation with a universal intrinsics. So I am not sure if this is suitable for merging, but if some could help and guide me to do that I will really appreciate that.

### CODE REVIEW

**Comment by rgommers** [comment]
> I have noticed a recent tendency of replacing newly added SIMD intrinsics of existing implementation with a universal intrinsics. So I am not sure if this is suitable for merging, but if some could help and guide me to do that I will really appreciate that.

Thanks for asking. It looks like that'd be the way to go here. Cc'ing @seiko2plus and @Qiyu8 as the universal intrinsics and Neon experts.

File: numpy/core/src/umath/simd.inc.src [hunk 1]
```diff
@@ -3697,4 +3736,143 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+#ifdef NPY_HAVE_NEON
+/**begin repeat
+ *  #type = npy_float, npy_double#
+ *  #TYPE = FLOAT, DOUBLE#
+ *  #scalarf = npy_sqrtf, npy_sqrt#
+ *  #Nvtype128 = float32x4_t, float64x2_t#
+ *  #Nvtype64 = float32x2_t, float64x1_t#
+ *  #Nvpre = v, v#
+ *  #Nvsuf = f32, f64#
+ *  #double = 0, 1#
```
Comment (eric-wieser) [inline numpy/core/src/umath/simd.inc.src:3748]: This seems not to be used

Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src]: unused `#double = 0, 1 ` deleted

File: numpy/core/src/umath/simd.inc.src [hunk 2]
```diff
@@ -3697,4 +3736,142 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+    else {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype128@ a = @Nvpre@ld1q_@Nvsuf@(&ip1[i]);
+            @Nvtype128@ b = @Nvpre@ld1q_@Nvsuf@(&ip2[i]);
+            @Nvtype128@ c = @Nvpre@@VOP@q_@Nvsuf@(a, b);
+            @Nvpre@st1q_@Nvsuf@(&op[i], c);
+        }
+    }
+
+#elif defined __aarch32__
```
Comment (Qiyu8) [inline numpy/core/src/umath/simd.inc.src:3782]: `defined __aarch32__` is redundant

File: numpy/core/src/umath/simd.inc.src [hunk 3]
```diff
@@ -3697,4 +3736,142 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+
+#endif
+    LOOP_BLOCKED_END {
+        op[i] = ip1[i] @OP@ ip2[i];
+    }
+}
+
+
+static void
+neon_binary_scalar1_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2, npy_intp n)
```
Comment (Qiyu8) [inline numpy/core/src/umath/simd.inc.src:3811]: scalar1 and scalar2 is too similar, can you extract a public function?

File: numpy/core/src/umath/simd.inc.src [hunk 4]
```diff
@@ -3697,4 +3736,142 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+* # kind = add, subtract, multiply, divide#
+* # OP = +, -, *, /#
+* # VOP = add, sub, mul, div#
+*/
+
+static void
+neon_binary_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2, npy_intp n)
+{
+#ifdef  __aarch64__
+    const npy_intp vector_size_bytes = 16;
```
Comment (Qiyu8) [inline numpy/core/src/umath/simd.inc.src:3762]: Can you use the common vector_size_bytes that was defined above? I suggest a macro define.

**Review by Qiyu8 — CHANGES_REQUESTED** [review]

File: numpy/core/src/umath/simd.inc.src (hunk 2 above)
Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src]: replaced with `#else`

File: numpy/core/src/umath/simd.inc.src (hunk 3 above)
Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src]: used macros to combine both into one.

File: numpy/core/src/umath/simd.inc.src (hunk 4 above)
Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src]: defined macro vector_size_bytes=16 (8) universally for __aarch64__ (__aarch32__) at the very beginning.

File: numpy/core/src/umath/simd.inc.src [hunk 5]
```diff
@@ -3697,4 +3739,113 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
 #endif /* NPY_HAVE_SSE2_INTRINSICS */
 
+#ifdef NPY_HAVE_NEON
+/**begin repeat
+ *  #type = npy_float, npy_double#
+ *  #TYPE = FLOAT, DOUBLE#
+ *  #scalarf = npy_sqrtf, npy_sqrt#
+ *  #Nvtype128 = float32x4_t, float64x2_t#
+ *  #Nvtype64 = float32x2_t, float64x1_t#
+ *  #Nvpre = v, v#
```
Comment (eric-wieser) [inline numpy/core/src/umath/simd.inc.src:3749]: This seems a bit silly, why not just use `v` below?

File: numpy/core/src/umath/simd.inc.src [hunk 6]
```diff
@@ -3697,4 +3739,113 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+    }
+    else {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype128@ a = @Nvpre@ld1q_@Nvsuf@(&ip1[i]);
+            @Nvtype128@ b = @Nvpre@ld1q_@Nvsuf@(&ip2[i]);
+            @Nvtype128@ c = @Nvpre@@VOP@q_@Nvsuf@(a, b);
+            @Nvpre@st1q_@Nvsuf@(&op[i], c);
+        }
+    }
+
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:3783]:
```suggestion
LOOP_BLOCK_ALIGN_VAR(op, @type@, NPY_SIMD_WIDTH)
        op[i] = ip1[i] @OP@ ip2[i];
    /* lots of specializations, to squeeze out max performance */
    if (ip1 == ip2) {
        LOOP_BLOCKED(@type@, NPY_SIMD_WIDTH) {
            npyv_@sfx@ a = npyv_load_@sfx@(&ip1[i]);
            npyv_@sfx@ c = npyv_@VOP@_@sfx@(a, a);
            npyv_store_@sfx@(&op[i], c);
        }
    }
    else {
        LOOP_BLOCKED(@type@, NPY_SIMD_WIDTH) {
            npyv_@sfx@ a = npyv_load_@sfx@(&ip1[i]);
            npyv_@sfx@ b = npyv_load_@sfx@(&ip2[i]);
            npyv_@sfx@ c = npyv_@VOP@_@sfx@(a, b);
            npyv_store_@sfx@(&op[i], c);
        }
    }
```

File: numpy/core/src/umath/simd.inc.src [hunk 7]
```diff
@@ -3697,4 +3739,113 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+    else {
+        LOOP_BLOCKED(@type@, vector_size_bytes) {
+            @Nvtype64@ a = @Nvpre@ld1_@Nvsuf@(&ip1[i]);
+            @Nvtype64@ b = @Nvpre@ld1_@Nvsuf@(&ip2[i]);
+            @Nvtype64@ c = @Nvpre@@VOP@_@Nvsuf@(a, b);
+            @Nvpre@st1_@Nvsuf@(&op[i], c);
+        }
+    }
+
+#endif
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:3804]:
```suggestion
```

File: numpy/core/src/umath/simd.inc.src [hunk 8]
```diff
@@ -3697,4 +3739,113 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+    LOOP_BLOCKED_END {
+        op[i] = ip1[@ip1ind@] @OP@ ip2[@ip2ind@];
+    }
+}
+/**end repeat2**/
+/**end repeat1**/
+/**end repeat**/
+
+#endif /* NPY_HAVE_NEON */
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:3850]:
```suggestion
    const npyv_@sfx@ v@scalar_loc@ = npyv_setall_@sfx@(ip@scalar_loc@);
    LOOP_BLOCK_ALIGN_VAR(op, @type@, NPY_SIMD_WIDTH)
    op[i] = ip1[@ip1ind@] @OP@ ip2[@ip2ind@];
    LOOP_BLOCKED(@type@, NPY_SIMD_WIDTH) {
        npyv_@sfx@ v@vector_loc@ = npyv_load_@sfx@(&ip@vector_loc@[i]);
        npyv_@sfx@ v3 = npyv_@VOP@_@sfx@(v1, v2);
        npyv_store_@sfx@(&op[i], v3);
    }
    LOOP_BLOCKED_END {
        op[i] = ip1[@ip1ind@] @OP@ ip2[@ip2ind@];
    }
}
/**end repeat2**/
/**end repeat1**/
#endif /* NPY_SIMD@CHK@ */
/**end repeat**/

```

Comment (seiko2plus) [reply numpy/core/src/umath/simd.inc.src:3850]: please separate them into two absolute functions

**Review by seiko2plus — CHANGES_REQUESTED** [review]
Thank you for your contribution.
We are now at the beginning of a new phase called "the universal intrinsics" big name but too easy to use.
Maybe it will be kinda hard to remove X86 SIMD code but at least the new code should be written via the new interface.
We don't have documentation yet but I guess you can still [discover it.](https://github.com/numpy/numpy/tree/master/numpy/core/src/common/simd)
The following changes should enable SIMD on both ARM(7/8) and PPC64LE.
**Note** that on ARM7 there's no kind of support for double-precision,
also, the changes are made on the fly so I may committed some mistakes.

File: numpy/core/src/umath/simd.inc.src [hunk 9]
```diff
@@ -553,6 +555,18 @@ static void
 sse2_binary_scalar2_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2,
                                   npy_intp n);
 
+#elif @VECTOR@
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:558]:
```suggestion
#elif !defined(NPY_HAVE_SSE2) && @VECTOR@
```
That should fix the CI error.

File: numpy/core/src/umath/simd.inc.src [hunk 10]
```diff
@@ -584,6 +598,25 @@ run_binary_simd_@kind@_@TYPE@(char **args, npy_intp const *dimensions, npy_intp
         sse2_binary_@kind@_@TYPE@(op, ip1, ip2, n);
         return 1;
     }
+#elif @VECTOR@
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:601]:
```suggestion
#elif !defined(NPY_HAVE_SSE2) && @VECTOR@
```
same here

File: numpy/core/src/umath/simd.inc.src [hunk 11]
```diff
@@ -3697,4 +3730,80 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
 
 #endif /* NPY_HAVE_SSE2_INTRINSICS */
 
+/**begin repeat
+ *  #type = npy_float, npy_double#
+ *  #TYPE = FLOAT, DOUBLE#
+ *  #sfx = f32, f64#
+ *  #CHK =    , _F64#
+ */
+#if NPY_SIMD@CHK@
```
Comment (seiko2plus) [inline numpy/core/src/umath/simd.inc.src:3739]:
```suggestion
#if !defined(NPY_HAVE_SSE2) && NPY_SIMD@CHK@
```
here too

File: numpy/core/src/umath/simd.inc.src (hunk 5 above)
Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src]: Silly indeed. Removed for universal intrinsics.

**Review by seiko2plus — APPROVED** [review]
LGTM, Thank you!.

File: numpy/core/src/umath/simd.inc.src [hunk 12]
```diff
@@ -3695,6 +3728,81 @@ sse2_@kind@_BOOL(@type@ * op, @type@ * ip, const npy_intp n)
...
+* # kind = add, subtract, multiply, divide#
+* # OP = +, -, *, /#
+* # VOP = add, sub, mul, div#
+*/
 
+static void
+simd_binary_@kind@_@TYPE@(@type@ * op, @type@ * ip1, @type@ * ip2, npy_intp n)
+{
+    LOOP_BLOCK_ALIGN_VAR(op, @type@, NPY_SIMD_WIDTH)
+        op[i] = ip1[i] @OP@ ip2[i];
```
Comment (eric-wieser) [inline numpy/core/src/umath/simd.inc.src:3751]: Can you add braces to match the `LOOP_BLOCKED_END`?

Comment (DumbMice) [reply numpy/core/src/umath/simd.inc.src:3752]: Braces added as required.

**Comment by DumbMice** [comment]
Thank you for your detailed guidance! @seiko2plus

**Comment by mattip** [comment]
CI failure does not seem to be related to this PR. 

Could you run benchmarks on the final version of this PR, and also compare the size of `_multiarray_umath*.so` before and after? Extra points for benchmarking x86 as well as arm64.

**Comment by DumbMice** [comment]
## Benchmark of the Last Commit

|      before      |    after      |  ratio | bench
|         ---      |    ---        | ---    | ---
|      196±0.8μs   |    117±0.7μs  |  0.59  | bench_ufunc.CustomInplace.time_double_add_temp
|    51.5±0.03μs   |  28.8±0.02μs  |  0.56  | bench_ufunc.CustomScalar.time_divide_scalar2_inplace(<class 'numpy.float32'>)
|        166±2μs   |     91.5±2μs  |  0.55  | bench_ufunc.CustomInplace.time_double_add
|    24.0±0.07μs   |   13.2±0.4μs  |  0.55  | bench_ufunc.CustomScalar.time_add_scalar2(<class 'numpy.float64'>)
|    50.4±0.07μs   |  27.3±0.05μs  |  0.54  | bench_ufunc.CustomScalar.time_divide_scalar2(<class 'numpy.float32'>)
|        347±2μs   |    126±0.6μs  |  0.36  | bench_ufunc.CustomInplace.time_float_add_temp
|     24.4±0.6μs   |  8.50±0.07μs  |  0.35  | bench_ufunc.CustomScalar.time_add_scalar2(<class 'numpy.float32'>)
|      321±0.4μs   |   95.0±0.4μs  |  0.30  | bench_ufunc.CustomInplace.time_float_add

Results showed performance close to the initial commit.

## File Size Change
File : numpy/core/_multiarray_umath.cpython-37m-aarch64-linux-gnu.so
|   | before | after | change |
| --- | --- | --- | --- |
| size | 16528K | 16589K | +61K |


## On x86
Compare neon_intrinsics with master:
BENCHMARKS NOT SIGNIFICANTLY CHANGED.

### x86 System Info
Hardware: Lenovo T470P
Processor: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
OS: 18.04.1-Ubuntu

**Comment by mattip** [comment]
Looks good. I think we should put this in, as a first use case. Correct me if I am wrong: this PR does not yet use the dispatch mechanism, nor does it replace the x86 loops with universal simd ones. It does set the baseline loops for add, sub, mul, div on arm64, power, and s390x to ones that use appropriate intrinsics for float32 and float64 loops.

**Comment by seiko2plus** [comment]
@mattip, yes, you're right except we don't support z/architecture(s390x) yet.  The current universal intrinsics code is only enabled for non-x86 CPU features and under the domain of the baseline features('--cpu-baseline').

**Comment by mattip** [comment]
Thanks @DumbMice

**Pull request merged**